### PR TITLE
Fix http-redirect conformance

### PR DESCRIPTION
**Pull request by T-Fowl**
**Subsystem**
Client, HttpRedirect feature

**Motivation**
While MDN **incorrectly** states that rewriting redirected POST requests to use GET is against the specification, it is actually **allowed** in RFC7231, specifically for [301 Moved Permanently and 302 Found](https://tools.ietf.org/html/rfc7231#page-56).
Most notably chrome, and probably all major browsers, do this (i.e. when submitting a login form which then receives a 302 Found back to the authenticated page).

**Solution**
Updated the HttpRedirect feature with a configuration class which allows the user to explicitly allow the rewrites of POST to GET (defaults to false) as well as a maximum number of redirects to follow (default is [RFC2068's 5](https://tools.ietf.org/html/rfc2068#section-10.3)).

This now allows users to do the following:
```kotlin
install(HttpRedirect) {
    maximumRedirects = 23 // Follow up to 23 redirects
    rewritePostAsGet = true // Enable rewrite to GET
}
```

Something that needs to be discussed, in the event that the client is redirected more than the configured number of times, I was unsure as to what to actually do:
```kotlin
// TODO: Just return last call, return original (closed) call, or throw an exception?
if (redirects > feature.maximumRedirects) return call
```

### CODE REVIEW

**Comment by cy6erGn0m** [comment]
Could you please provide a test?

**Comment by T-Fowl** [comment]
Added some tests but have spent all afternoon trying to work out why some of them aren't working. Specifically ones that involve HttpRedirect throwing an exception when too many redirects are received. HttpSend currently does this (at 20) but I cannot manage to replicate this in HttpRedirect.  


The following is in HttpRedirect:
```kotlin
if (redirects >= feature.maximumRedirects)
    throw RedirectCountExceedException("Max redirect count ${feature.maximumRedirects} exceeded")
```

Then on line 62-76 of HttpSend:
```kotlin
var currentCall = sender.execute(context) // SendCountExceedException gets thrown here
var callChanged: Boolean

do {
    callChanged = false

    passInterceptors@ for (interceptor in feature.interceptors) {
        val transformed = interceptor(sender, currentCall) // will be thrown here
        if (transformed === currentCall) continue@passInterceptors

        currentCall = transformed
        callChanged = true
        break@passInterceptors
    }
} while (callChanged)
```

The exception thrown in HttpRedirect should flow up the call stack into the interceptor loop in HttpSend and cause the same effect as the SendCountExceedException exception in HttpSend. However this exception seems to be lost somewhere.

Tests requiring this exception to be thrown simply timeout.

The redirect loop in HttpRedirect exits, but the call does not seem to complete at all (let alone throwing)

However this exception is not having an effect on the pipeline execution during testing and is causing tests to timeout.

File: ktor-client/ktor-client-core/common/src/io/ktor/client/features/HttpRedirect.kt [hunk 1]
```diff
@@ -39,9 +70,18 @@ class HttpRedirect {
                     url.parameters.clear()
 
                     location?.let { url.takeFrom(it) }
+
+                    if (call.response.status.isRewritable()) {
+                        if (feature.rewritePostAsGet && call.request.method == HttpMethod.Post)
```
Comment (e5l) [inline ktor-client/ktor-client-core/common/src/io/ktor/client/features/HttpRedirect.kt]: Could you merge it on one `if`?

File: ktor-client/ktor-client-tests/common/test/io/ktor/client/tests/HttpRedirectTest.kt [hunk 2]
```diff
@@ -30,6 +31,33 @@ class HttpRedirectTest : ClientLoader() {
...
+    @Test
+    fun rewrittenRedirectTest(): Unit = clientTests {
+        config {
+            install(HttpRedirect) {
+                rewritePostAsGet = true
+            }
+        }
+
+        test { client ->
+            client.post<HttpResponse>("$TEST_URL_BASE/post-expecting-get-301").use {
```
Comment (e5l) [inline ktor-client/ktor-client-tests/common/test/io/ktor/client/tests/HttpRedirectTest.kt]: Could you rebase tests on master?

File: ktor-client/ktor-client-core/common/src/io/ktor/client/features/HttpRedirect.kt [hunk 3]
```diff
@@ -39,9 +70,18 @@ class HttpRedirect {
...
+                            method = HttpMethod.Get
+                    }
                 })
+                redirects++
 
                 if (!call.response.status.isRedirect()) return call
+
+                if (redirects >= feature.maximumRedirects)
+                    throw RedirectCountExceedException("Max redirect count ${feature.maximumRedirects} exceeded")
```
Comment (e5l) [inline ktor-client/ktor-client-core/common/src/io/ktor/client/features/HttpRedirect.kt]: It's quite ambiguous with a limit in the `HttpSend` feature - we throw 2 kinds of exceptions for similar reasons.
Could you tell me if you have an idea of how we can simplify the handling of such cases?

**Comment by e5l** [comment]
Hi @T-Fowl, thanks for the PR. It's almost good, except minor usability issues.

**Pull request closed without merging**